python/lsst/daf/butler/remote_butler/server/handlers/_external_query.py

```diff
@@ -64,16 +64,60 @@
 # Alias this function so we can mock it during unit tests.
 _timeout = asyncio.timeout
 
+_TContext = TypeVar("_TContext")
+
+
+class StreamingQuery(Protocol[_TContext]):
+    """Interface for queries that can return streaming results."""
+
+    def setup(self) -> AbstractContextManager[_TContext]:
+        """Context manager that sets up any resources used to execute the
+        query.
+        """
+
+    def execute(self, context: _TContext) -> Iterator[QueryExecuteResultData]:
+        """Execute the database query and and return pages of results.
+
+        Parameters
+        ----------
+        context : generic
+            Value returned by the call to ``setup()``.
+        """
+
+
+class _StreamQueryDriverExecute(StreamingQuery):
```

Comment:
I think you might need to repeat the `[_TContext]` generic parameterization here to avoid having it implicitly interpreted as `StreamingQuery[Any]`, but I'm not sure.